Preliminary macOS support

smcio opened the pull request:
Preliminary support for macOS.

There some few important differences between Linux and macOS machines in the context of firedancer's architecture and implementation, with some of the most important being huge page support and CPU pinning.

I've attempted to work around these problems, and I'll follow up with some more details (likely tomorrow), but recent Mac systems (i.e. M1 and M2) do not appear to offer a userland interface for allocation of memory using large page sizes. macOS running on Intel hardware does support huge pages, but seemingly 2MB only. I've attempted to work around these problems, such as by allowing the use of normal sized pages.

I'm putting this up as a draft for your review and for requesting changes whenever you've had time to take a look :slight_smile:

smcio commented:
Cheers @ripatel-jump, much appreciated. I'll respond & begin making changes for all your comments later today.

kbowers-jump commented on config/macos_clang_arm64.mk:
```diff
@@ -0,0 +1,23 @@
+# -falign-functions=32 -falign-jumps=32 -falign-labels=32 -falign-loops=32
+# -mbranch-cost=5
+
+CPPFLAGS+=-fomit-frame-pointer -march=native -mtune=skylake \
+	  -DFD_HAS_INT128=1 -DFD_HAS_DOUBLE=1 -DFD_HAS_ALLOCA=1 -DFD_HAS_X86=1
```
There's a lot of history around `-fomit-frame-pointer`.

x86 has always been severely register staved versus other architectures and it was really bad pre-x86_64 (the larger register file in x86_64 was as at least important as having 64-bit registers).

So, in the way, way back on Linux and x86, you'd use this compile flag in a optimized build to make an extra register available for code.  It was pretty dramatic improvement in real world performance to have an extra register when you only had O(6) general purpose-ish registers available (eax,ebx,ecx,edx,edi,esi) ... e.g. fewer spills to stack ("-ish" because the x86 ISA does have lots of hardcoded usages for these registers).

This flag wasn't a pure benefit though.  It made debugging binaries more difficult as it was more difficult for a debugger to inspect what was going on in the stack.

By comparison, non-x86 at the time typically had O(16 to 32) truly general purposes registers (so the potential benefit was lower) and the way non-Linux OS / ABIs / etc/ managed stack, handled function calls, etc could affect the cost (from irrelevant to impossible).

Since then, debuggers on Linux / binary file debugging symbol info / etc got more sophisticated.  This make the debugging visibility cost of this flag lower.  At the same time, x86_64 has a larger (but still not great) register file, the benefit is lower too.  And architectures and O/S might have evolved too.

The upshot of all this, `-fomit-frame-pointer` is somewhere between useful and moot on modern Linux / x86_64 (and, if somebody has some compelling reason to require or remove it on modern Linux / x86_64, totally open to it).

As for non-x86 and/or non-Linux, it is probably somewhere between annoying to mildly useful due to analogous considerations specific that architecture and OS (and/or there might be some other architecture specific flags in this vein).

The upshot being, `-fomit-frame-pointer` it certainly isn't a requirement to have and should be viewed much much as a platform specific flag.  It is debatable how useful it is on Linux / x86_64 these days.  And it is even more debatable on non-Linux and/or non-x86_64.

smcio replied:
@kbowers-jump  @ripatel-jump  Ah, see, the reason I've used x64 flags for building on macOS ARM (M2 in my case, but also M1) is because I am building using Rosetta. I can certainly work on making this build and work under macOS M1/M2 without Rosetta if you'd like.

kbowers-jump commented on config/macos_clang_arm64.mk:
```diff
@@ -0,0 +1,23 @@
+# ideally would do:
+#
+# -falign-functions=32 -falign-jumps=32 -falign-labels=32 -falign-loops=32
+# -mbranch-cost=5
```
This specific comment is architecture specific but there are likely analogous considerations for other architectures.

(x86 ISA use of variable width instructions is good for instruction footprint / L1I cache pressure but makes life hard for the instruction decode stages of the CPU pipeline ... can look at things like ARM or RISC-V, where there are instruction encoding options to make efficient storage in memory and fast / low area decoding and part of that is often having different and more restrictive alignment requirements for instructions baked in day 1).

smcio replied:
Will do, thanks.

kbowers-jump commented on config/macos_clang_x86_64.mk:
```diff
@@ -0,0 +1,25 @@
+# -falign-functions=32 -falign-jumps=32 -falign-labels=32 -falign-loops=32
+# -mbranch-cost=5
+
+CPPFLAGS+=-fomit-frame-pointer -march=haswell -mtune=skylake -mfpmath=sse \
+	  -DFD_HAS_INT128=1 -DFD_HAS_DOUBLE=1 -DFD_HAS_ALLOCA=1 -DFD_HAS_X86=1 -DFD_HAS_SSE=1 -DFD_HAS_AVX=1
```
Not sure but, if it is an x86 CPU MAC, I'd be mildly surprised if AVX support wasn't there.

That is, I'd guess the CPU itself does have it physically in that the level of AVX-ness required for Firedancer as is today was widespread by the mid 2010s (though having a custom SKU made for large enough volume customer isn't completely unheard of).

Likewise, I'd guess the OS does have it logically because even PowerPC CPU used in the pre-x86 CPU MAC era had features like Altivec for vectorization (though it is possible that the OS itself doesn't allow user space applications to play with AVX features for power or context switching performance or vendor flexibility or some other legacy compat reasons).

smcio replied:
'Recent' x64/x86 Macs have AVX support, but ARM Macs do not have AVX support, so I've omitted AVX flags in the mac ARM Makefile, but included them in the x64 one.

I could find out when Mac (x64) added AVX support and document that. I figured that in the x64 case, use of 'older' Macs without AVX would be pretty unlikely.

smcio commented:
Cheers @ripatel-jump. I'll push up the changes per the discussion in this thread tomorrow.

smcio commented on config/macos_clang_arm64.mk:
```diff
@@ -0,0 +1,23 @@
+FD_HAS_INT128:=1
+FD_HAS_DOUBLE:=1
+FD_HAS_ALLOCA:=1
+FD_HAS_X86:=1
```
As above, the reason for using x64/x86 flags in the macOS ARM Makefile is because I was testing this on my macOS M2 machine using Rosetta. Also as above, I can work on making this build & run correctly without Rosetta, though this will seemingly become https://github.com/firedancer-io/firedancer/issues/99

smcio commented on config/everything.mk:
```diff
@@ -174,7 +174,7 @@ $(OBJDIR)/obj/%.d : src/%.c
 	#######################################################################
 	$(MKDIR) $(dir $@) && \
-$(CC) $(CPPFLAGS) $(CFLAGS) -M -MP $< -o $@.tmp && \
+$(CC) $(CPPFLAGS) $(CFLAGS) -M $< -o $@.tmp && \
```
I've reverted this change. I removed the `-MP` flag at some point, but why exactly escapes me now, since this flag appears fine for macOS on both x64 and M1/M2 🙂

smcio commented:
I've addressed the comments above by pushing some new changes, and I've marked this as Ready for Review so that you guys can take a better look and let me know where I've made oversights given my lack of through knowledge Re: the codebase so far. I've worked around several API differences (incl. missing APIs), including working around limited support for large page sizes on macOS x64, and seemingly no userland large pages support on M1 & M2. 

I have the tests passing and the implementation running, but some or many of the way I've worked around the several API differences may seem/be a bit hacky, so I'd be interested in improving these if better and cleaner solutions occur to you. Let me know your thoughts - I'll be around, incl. on Discord.

Cheers,
Shaun

jumpsiegel commented:
Could this get rebased onto main?

jumpsiegel commented:
ahhh... never mind.. :)
